default/scripting/policies/COLONIALISM.focs.txt

```diff
@@ -44,14 +44,11 @@ Policy
                         Planet
                         OwnedBy empire = Source.Owner
                         Focus type = "FOCUS_INDUSTRY"
-                        ResourceSupplyConnected empire = Source.Owner condition = And [
-                            Not InSystem id = RootCandidate.SystemID
-                            Planet
-                            Species
-                            OwnedBy empire = Source.Owner
-                            Focus type = "FOCUS_INDUSTRY"
-                            WithinStarlaneJumps jumps = 3 condition = RootCandidate
-                        ]
+                        // TODO: this should actually be: a jump route not longer than 3 jumps where all hops are supplied
+                        //       currently one reaps benefits through blockades as long as both endpoints are in the same supply
+                        //       this would need a new condition able to exclude planets/ a filter based on the hops
+                        WithinStarlaneJumps jumps = 3 condition = RootCandidate
```

Comment:
If I understand correctly, `RootCandidate` on lines 43, 50, and 51 is the same object as `LocalCandidate` on line 42, since there is no other enclosing condition being evaluated, so the root candidate of the context is initially nothing and is set to the local candidate as soon as something is being matched by the condition.

The `LocalCandidate` or `RootCandidate` on line 42 is a planet that was matched by `condition = And [ ...` on line 53, and is a planet that `Target` is taking industry from.

Reply:
So, as written with these changes, `WithinStarlaneLumps` on line 50 is seemingly checking if an object is within 3 jumps of itself, and similarly for `ResourceSupplyconnected` on line 51.

Reply:
Whereas the previous version was checking if there is another industry-focused populated planet owned by the source's owner that is not in the `RootCandidate`'s system but is within 3 jumps that is resource supply connected to the `RootCandidate`

Reply:
> If I understand correctly, `RootCandidate` on lines 43, 50, and 51 is the same object as `LocalCandidate` on line 42, 
yes, well that is the idea. the topmost LocalCandidate is the current object out of the object set returned from the Statistic condition (over which must be aggregated).

in my understanding the RootCandidate luckily falls back to the first LocalCandidate - which is lucky/maybe a bit brittle

i cant follow your reasoning on what the code does. maybe you confused what the candidate objects are in the value valueref? those are planets where industry is taken from (which i call an industry source for lack of a better word).

with these changes, the count condition counts the number of planets which are resource connected to the candidate (which is an industry source) and within 3 jumps from it (and are legitimate industry sinks). 

the previous version was counting planets which are legitimate industry sinks which are supply connected and within 3 jumps of other industry sinks... which is meaningless?

Reply:
> in my understanding the RootCandidate luckily falls back to the first LocalCandidate - which is lucky/maybe a bit brittle

"falls back" is perhaps misleading... The design is that, when evaluating a condition, if there is no root candidate in the context, then the local candidate is the root candidate. For example, a top level ValueRef that has a condition in it... The ValueRef has no local or root candidates, since it's not a condition. When evaluating the contained condition, every candidate object is also a root candidate, and the passed in candidate is set to both the local and root candidate fields in the constructed ScriptingContext for each during evaluation.